File: backend-spring-boot/src/main/java/no/vicx/backend/calculator/CalculatorSecurityService.java
package no.vicx.backend.calculator;

import no.vicx.database.calculator.CalculatorRepository;
import org.springframework.lang.NonNull;
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken;
import org.springframework.stereotype.Service;

import java.util.Collection;
import java.util.Objects;
import java.util.stream.Collectors;

@Service
public record CalculatorSecurityService(CalculatorRepository calculatorRepository) {

    public boolean isAllowedToDelete(
            @NonNull Collection<Long> ids,
            @NonNull JwtAuthenticationToken authentication) {
        if (ids.isEmpty()) {
            return true; // let validation handle this
        }

        var nonNullIds = ids.stream()
                .filter(Objects::nonNull)
                .collect(Collectors.toSet());

        if (nonNullIds.isEmpty()) {
            return true; // let validation handle this
        }

        var idsFromDatabase = calculatorRepository.findAllIdsByUsername(authentication.getName());
        return idsFromDatabase.containsAll(nonNullIds);
    }
}

File: backend-spring-boot/src/main/kotlin/no/vicx/backend/calculator/CalculatorSecurityService.kt
package no.vicx.backend.calculator

import no.vicx.database.calculator.CalculatorRepository
import org.springframework.security.core.Authentication
import org.springframework.stereotype.Service


@Service
class CalculatorSecurityService(
    private val calculatorRepository: CalculatorRepository
) {
    fun isAllowedToDelete(
        idsToDelete: Collection<Long>,
        authentication: Authentication
    ): Boolean {
        if (idsToDelete.isEmpty()) return true // let validation handle this

        val idsFromDatabase = calculatorRepository.findAllIdsByUsername(authentication.name)
        return idsFromDatabase.containsAll(idsToDelete)
    }
}
